abci: Application should return errors errors and nilable response objects

tychoish opened the pull request:
Please add a description of the changes that this PR introduces and the files that
are the most critical to review.

If this PR fixes an open Issue, please include "Closes #XXX" (where "XXX" is the Issue number) 
so that GitHub will automatically close the Issue when this PR is merged.

thanethomson reviewed (commented):
I think we may need some synchronous discussion about this sometime. This seems like quite a fundamental change to ABCI(++), and potentially has protocol- and specification-level implications.

For instance, what does Tendermint do with errors coming from the application?

And how is this handled when interacting with the app via the Tendermint Socket Protocol? If these errors are meaningful for Tendermint, the ABCI response protos would need to be augmented to be able to convey these errors back to Tendermint.

tychoish commented:
> For instance, what does Tendermint do with errors coming from the application?
> 
> And how is this handled when interacting with the app via the Tendermint Socket Protocol? If these errors are meaningful for Tendermint, the ABCI response protos would need to be augmented to be able to convey these errors back to Tendermint.

I think its actually far less disruptive than you think: the socket client and server have always propogated errors from the service layer (you'll note that at that layer none of the function signatures changed.) now we're letting the application add additional errors. (Previously applications could panic, which of course they still can.) so from all areas of the code above the abci package, there's no change. 

The change is mostly impacting applications in the sense that their interface change changes a bit more (e.g. in addition to all of the method names and types changing for ABCI++ now they have a context, and different return types.) For about 50% of the interface, these are new anyway, so the addition of this patch doesn't represent a change on top of the existing changes since the last release. For the second half of the interface, it's *probably* a good thing that we're breaking these interfaces because we want application developers to go in and look at the way the ABCI methods are implemented because we took out the mutex (which doesn't break compile, of course) so while it represents a bit of work, it's work that they should be doing anyway.

creachadair commented on abci/types/application.go:
```diff
@@ -5,34 +5,32 @@ import "context"
+	ListSnapshots(context.Context, RequestListSnapshots) (*ResponseListSnapshots, error)                // List available snapshots
+	OfferSnapshot(context.Context, RequestOfferSnapshot) (*ResponseOfferSnapshot, error)                // Offer a snapshot to the application
+	LoadSnapshotChunk(context.Context, RequestLoadSnapshotChunk) (*ResponseLoadSnapshotChunk, error)    // Load a snapshot chunk
+	ApplySnapshotChunk(context.Context, RequestApplySnapshotChunk) (*ResponseApplySnapshotChunk, error) // Apply awaq  a shapshot chunk
```
```suggestion
	ApplySnapshotChunk(context.Context, RequestApplySnapshotChunk) (*ResponseApplySnapshotChunk, error) // Apply a shapshot chunk
```

tychoish commented:
I also included, in the last commit, some (rough) cleanup to the spec in the relevant sections.

thanethomson commented on abci/types/application.go:
```diff
@@ -5,34 +5,32 @@ import "context"
 	// Info/Query Connection
-	Info(context.Context, RequestInfo) ResponseInfo    // Return application info
-	Query(context.Context, RequestQuery) ResponseQuery // Query for state
+	Info(context.Context, RequestInfo) (*ResponseInfo, error)    // Return application info
```
What's the benefit of having the responses be references to structs, but the requests not?

tychoish replied:
One (I?) want return values to be nil, so in the error case you can just return nil. For the argument struct--in most cases--you can construct the inline to the function argument, and it's not a big deal. I'm not worried about the copy (struct values) or the heap allocation (pointers), particularly, just the semantics. 

One of the big changes, (and I think useful parallels) at this point is that the Client interface and the application interface are now the same. (basically).

creachadair replied:
Adding on to @tychoish's answer: These types are mostly protobuf generated types, and the generated methods are all on the pointer anyway. Although there's a lot of existing use where we pass these around by value, that's generally _not_ recommended for protobuf types. Apart from the costs of shipping around potentially very large values (proto structs are big and tend to grow over time), passing by value forces us to do some shenanigans when we want to handle them (e.g., adding/removing pointers).

Ideally I'd like to fix _all_ the places we pass protobuf types by value, but these are a good start (and since it's new API surface, it won't cause any major issues).

thanethomson replied:
I suppose my question is, why not just pass the requests by reference then too?

creachadair replied:
> I suppose my question is, why not just pass the requests by reference then too?

I'm in favour of it. 🙂

thanethomson replied:
> I'm in favour of it. :slightly_smiling_face:

Same.

tychoish replied:
we can do that in another PR, but I think we should change the client API in that case.

creachadair commented:
> For instance, what does Tendermint do with errors coming from the application?
> 
> And how is this handled when interacting with the app via the Tendermint Socket Protocol? If these errors are meaningful for Tendermint, the ABCI response protos would need to be augmented to be able to convey these errors back to Tendermint.

I'm happy to talk more about it either way, but I don't think this will have any negative effects. Right now, applications can't report errors anyway, so we already have to check for response codes in the application result. We will continue to do that.

As @tychoish points out, it will require updating some method signatures, but anyone adopting ABCI++ is already going to have to do that anyway.
